File: my-folder/0424-longest-repeating-character-replacement/solution.py
from collections import defaultdict
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        L = 0  
        counter = defaultdict(int)
        longest_substring = 0
        maxFreq = 0
        for R in range(len(s)):
            counter[s[R]] += 1
            maxFreq = max(maxFreq, counter[s[R]])
            while(R - L + 1 - maxFreq > k):
                counter[s[L]] -= 1
                L += 1
            longest_substring = max(R - L + 1, longest_substring)
      
        return longest_substring

        

            
            

            


        

File: my-folder/0424-longest-repeating-character-replacement/solution.java
class Solution {
    public int characterReplacement(String s, int k) {
        int[] cntMap = new int[26];
        int maxFreq = 0;
        int left = 0;
        int lcr = 0;
        for(int right = 0; right < s.length(); right++){
            char curr = s.charAt(right);
            cntMap[curr - 'A'] += 1;
            maxFreq = Math.max(maxFreq, cntMap[curr - 'A']);
            while(right - left + 1 - maxFreq > k){
                char prev = s.charAt(left);
                cntMap[prev- 'A'] -= 1;
                left++;
                for(int i = 0; i < 26; i++){
                    maxFreq = Math.max(maxFreq, cntMap[i]);
                }
            }

            lcr = Math.max(right - left + 1, lcr);
        }
        return lcr;
    }
}
// window = {a : 2, b : 2} maxfreq = 2; 4 - 2 update lcr = to window length and return it

// ABAB 
// k = 2
